pyat/at/errors/error_def.py

```diff
@@ -0,0 +1,449 @@
+          flags. Default: :py:obj:`True`
+        shiftx (bool):      enable horizontal shift errors. Default: *all*
+        shifty (bool):      enable vertical shift errors. Default: *all*
+        tilt (bool):        enable tilt errors. Default: *all*
+        pitch (bool):       enable pitch errors. Default: *all*
+        yaw (bool):         enable yaw errors. Default: *all*
+        PolynomB (bool):    enable polynomial errors. Default: *all*
+        IndexB (Optional[int]): restrict the polynomial errors to the
+          specified index. Default: :py:obj:`None` meaning the whole polynom
+        PolynomA (bool):    enable polynomial errors. Default: *all*
+        IndexA (Optional[int]): restrict the skew polynomial error to the
+          specified index. Default: :py:obj:`None` meaning the whole polynom
+
+    Examples:
+        >>> ringerr = enable_errors(ring)
+
+        All errors are enabled in *ringerr*
+
+        >>> ringerr = enable_errors(all=False, ShiftErr=True)
+
+        Only magnet shift errors are active
+
+        >>> ringerr = enable_errors(all=False)
+
+        No magnet errors are enabled. Only monitor errors will be active.
+
+    See also:
+        :py:func:`assign_errors`, :py:func:`get_optics_err`
+    """
+    if not ring.has_errors:
+        ring = _apply_field_errors(ring, **kwargs)
+        ring = _apply_alignment_errors(ring, **kwargs)
+        ring._has_errors = True
+    return ring
+
+
+def has_errors(ring):
+    """Get the error status of a lattice"""
+    return getattr(ring, '_has_errors', False)
+
+
+def get_mean_std_err(ring: Lattice, key, attr, index=0):
+    vals = [np.atleast_1d(getattr(e, attr, 0.0))[index]
+            for e in ring.get_elements(key)]
+    return np.mean(vals), np.std(vals)
+
+
+def _sort_flags(kwargs):
+    errargs = {}
+    for key in _SEL_ARGS:
+        if key in kwargs:
+            errargs[key] = kwargs.pop(key)
+    return kwargs, errargs
+
+
+def find_orbit_err(ring: Lattice, refpts: Refpts = None, **kwargs):
+    """Find the closed orbit if a lattice with errors
+
+    :py:func:`find_orbit_err` enables the selected errors, finds the closed
+    orbit and modifies the result according to monitor errors.
+    The *ring* :py:class:`~.Lattice` is not modified.
+
+    Args:
+        ring:       Lattice description
+        refpts:
+
+    The *all*, *ShiftErr*, *RotationErr*, *PolynomBErr*, *PolynomBIndex*,
+    *PolynomAErr*, *PolynomAIndex* keywords are defined as in
+    :py:func:`enable_errors`
+
+    All other keywords are forwarded to :py:func:`.find_orbit`
+
+    See also:
+        :py:func:`assign_errors`, :py:func:`.find_orbit`,
+        :py:func:`get_optics_err`, :py:func:`lattice_pass_err`
+    """
+    kwargs, errargs = _sort_flags(kwargs)
+    ring = enable_errors(ring, **errargs)
+    orbit0, orbit = find_orbit(ring, refpts=refpts, **kwargs)
+    _apply_bpm_orbit_error(ring, refpts, orbit)
```

Comment:
`orbit0` does not need to be modified?

Reply:
the first element of the ring may be a BPM

Reply:
This is true, we could add the BPM errors to `orbit0` in case BPM errors are set for the first ring element, and in a similar way affect the closed orbit output of `get_optics`.

Reply:
I think you are mixing 2 things: orbit and BPM readings. `orbit0` is the orbit, it is independent of BPM errors and must stay so because it's the input for subsequent computations (matrix, optics, envelope). Only `orbit` is affected.

The 1st element may be a BPM, and if you select it in `refpts`, you get its reading as for any other element in orbit.

Ideally I would have added another output `orbit_reading` to keep it separate from `orbit`…